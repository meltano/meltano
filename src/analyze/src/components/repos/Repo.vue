<template>
  <div class="container">
    <div class="columns">
        <aside class="fixed-sidebar column is-one-quarter menu has-background-white-bis">
          <div class="level">
            <div class="level-left">
              <div class="field has-addons">
                <div class="control">
                  <a
                    href="#"
                    class="button is-small"
                    :class="{'is-loading': loadingValidation}"
                    @click="lint">Lint</a>
                </div>
                <div class="control">
                  <a
                    href="#"
                    class="button is-small"
                    :class="{'is-loading': loadingUpdate}"
                    @click="sync">Sync</a>
                </div>
              </div>
            </div>
            <div class="level-right">
              <span class="tag is-success pull-right" v-if="passedValidation">Passed!</span>
              <span class="tag is-warning" v-if="!validated">Unvalidated</span>
              <span class="tag is-danger" v-if="hasError">Errors</span>
            </div>
          </div>
          <template v-if="hasError">
            <nav class="panel">
              <!-- eslint-disable-next-line vue/require-v-for-key -->
              <div class="panel-block has-background-white" v-for="err in errors">
                <ul>
                  <li class="level">
                    <div class="tags has-addons">
                      <span class="tag is-info">?</span>
                      <span class="tag">{{err.file_name}}</span>
                    </div>
                  </li>
                  <li class="error-desc-cont">
                    <code class="error-desc">{{err.message}}</code>
                  </li>
                </ul>
              </div>
            </nav>
          </template>
          <template v-for="(value, key) in files">

            <!-- eslint-disable-next-line vue/require-v-for-key -->
            <p class="menu-label">
              <a href="#">{{value.label}}</a>
            </p>
            <!-- eslint-disable-next-line vue/require-v-for-key -->
            <ul class="menu-list">

              <template v-if="!value.items.length">
                <li>
                  <a><small><em>No {{value.label}}</em></small></a>
                </li>
              </template>

              <template v-if="value.items.length">
                <li v-for="file in value.items" :key="file.abs">
                  <div class="columns">
                    <div class="column">
                      <a :class="{'is-active': isActive(file)}"
                          @click.prevent='getFile(file)'>
                        {{file.name}}
                      </a>
                    </div>
                    <div v-if='isDeepRoutable(key)' class='column is-one-fifth'>
                      <router-link :to="getDeepRoute(key, file)"
                                    class="button is-secondary is-light is-pulled-right">
                        <span class="icon is-small">
                          <i class="fas fa-bold">*</i>
                        </span>
                      </router-link>
                    </div>
                  </div>

                </li>
              </template>

            </ul>

          </template>
        </aside>
      <div class="column" v-if="!activeView.populated">
        <div
          class="empty-state
          content
          has-background-white
          has-text-centered
          is-size-4
          is-uppercase">
          Select a file
        </div>
      </div>
      <div class="column" v-if="hasMarkdown">
        <div class="content has-background-white" v-html="activeView.file"></div>
      </div>
      <div class="column is-paddingless code-container" v-else-if="hasCode">
        <div class="content has-background-white">
          <pre>{{activeView.file | pretty}}</pre>
        </div>
      </div>
    </div>
  </div>
</template>
<script>
import { mapState, mapGetters } from 'vuex';
import capitalize from '@/filters/capitalize';
import pretty from '@/filters/pretty';

export default {
  name: 'Repo',
  created() {
    this.getRepo();
    this.sync();
  },
  filters: {
    pretty,
  },
  computed: {
    ...mapGetters('repos', [
      'hasError',
      'passedValidation',
      'hasMarkdown',
      'hasCode',
    ]),
    ...mapState('repos', [
      'files',
      'activeView',
      'validated',
      'loadingValidation',
      'loadingUpdate',
      'errors',
    ]),
  },
  methods: {
    getRepo() {
      this.$store.dispatch('repos/getRepo');
    },
    isActive(f) {
      return f.id === this.activeView.id;
    },
    isDeepRoutable(type) {
      return type === 'dashboards' || type === 'reports';
    },
    getDeepRoute(key, file) {
      const name = capitalize(key).slice(0, -1);
      const params = { slug: file.slug };
      if (file.model && file.design) {
        params.model = file.model;
        params.design = file.design;
      }
      return { name, params };
    },
    getFile(file) {
      this.$store.dispatch('repos/getFile', file);
    },
    lint() {
      this.$store.dispatch('repos/lint');
    },
    sync() {
      this.$store.dispatch('repos/sync');
    },
  },
};
</script>
<style lang="scss" scoped>
.content {
  padding: 10px;
  position: fixed;
  top: 52px;
  left: 430px;
  right: 0;
  bottom: 0;
  overflow: scroll;
  z-index: 1;
}

.code-container {
  // to overcome the important on paddingless
  padding-left: 1rem !important;
}

.error-desc-cont {
  .tag {
    margin-bottom: 0.5rem;
  }
  .error-desc {
    display: block;
  }
}

.empty-state {
  padding-top: 200px;
}

</style>
