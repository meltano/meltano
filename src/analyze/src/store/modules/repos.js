import repoApi from '../../api/repo';

const state = {
  activeView: { is_markdown: false, file: '', populated: false },
  loadingValidation: false,
  loadingUpdate: false,
  models: [],
  validated: false,
  navbarClicked: false,
  errors: [],
  files: {},
};

const getters = {

  hasMarkdown() {
    return state.activeView.populated && state.activeView.is_markdown;
  },

  urlForModelDesign: () => (model, design) => `/analyze/${model}/${design}`,

  hasCode() {
    return state.activeView.populated && !state.activeView.is_markdown;
  },

  hasError() {
    return state.errors && state.errors.length;
  },

  hasFiles() {
    return Object.hasOwnProperty.call(state.files, 'topics') && state.files.topics.items;
  },

  passedValidation() {
    return state.validated && state.errors && !state.errors.length;
  },

};

const actions = {
  getRepo({ commit }) {
    repoApi.index()
      .then((response) => {
        const files = response.data;
        commit('setValidatedState', response.data);
        state.loadingValidation = false;
        commit('setRepoFiles', { files });
      });
  },

  getFile({ commit }, file) {
    repoApi.file(file.id)
      .then((response) => {
        commit('setCurrentFileTable', response.data);
      });
  },

  lint({ commit }) {
    state.loadingValidation = true;
    repoApi
      .lint()
      .then((response) => {
        commit('setValidatedState', response.data);
        state.loadingValidation = false;
      })
      .catch(() => {
        state.loadingValidation = false;
      });
  },

  sync({ commit, dispatch }) {
    state.loadingUpdate = true;
    repoApi
      .sync()
      .then((response) => {
        dispatch('getModels');
        commit('setValidatedState', response.data);
        state.loadingUpdate = false;
      })
      .catch(() => {
        state.loadingUpdate = false;
      });
  },

  getModels({ commit }) {
    repoApi.models()
      .then((response) => {
        commit('setModels', response.data);
      });
  },

  navbarHideDropdown({ commit }) {
    commit('setHiddenDropdown');
  },
};

const mutations = {
  setHiddenDropdown() {
    state.navbarClicked = true;
    setTimeout(() => {
      state.navbarClicked = false;
    }, 500);
  },

  setModels(_, models) {
    state.models = models;
  },

  setRepoFiles(_, { files }) {
    state.files = files;
  },

  setCurrentFileTable(_, file) {
    state.activeView = file;
  },

  setValidatedState(_, validated) {
    state.errors = [];
    state.validated = true;
    // validation failed, so there will be errors
    if (!validated.result) {
      state.errors = validated.errors;
    }
  },

};

export default {
  namespaced: true,
  state,
  getters,
  actions,
  mutations,
};
