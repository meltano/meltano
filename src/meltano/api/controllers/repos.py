import json
import os
from os.path import join
from pathlib import Path

import markdown
from flask import jsonify, request
from flask_principal import Need
from meltano.api.api_blueprint import APIBlueprint
from meltano.api.json import freeze_keys
from meltano.api.security.auth import permit
from meltano.api.security.resource_filter import NameFilterMixin, ResourceFilter
from meltano.core.compiler.project_compiler import ProjectCompiler
from meltano.core.m5o.m5o_collection_parser import (
    M5oCollectionParser,
    M5oCollectionParserTypes,
)
from meltano.core.m5o.m5o_file_parser import (
    MeltanoAnalysisFileParser,
    MeltanoAnalysisFileParserError,
    MeltanoAnalysisMissingTopicFilesError,
)
from meltano.core.project import Project

from .repos_helper import ReposHelper

reposBP = APIBlueprint("repos", __name__)


class ReportIndexFilter(NameFilterMixin, ResourceFilter):
    def __init__(self, *args):
        super().__init__(*args)

        self.needs(self.design_need)

    def design_need(self, permission_type, report):
        return Need("view:design", report["design"])


class M5ocFilter(ResourceFilter):
    def filter_designs(self, designs):
        return (
            ResourceFilter().needs(self.design_need).filter_all("view:design", designs)
        )

    def design_need(self, permission_type, design_name):
        return Need(permission_type, design_name)

    def scope(self, permission_type, m5oc):
        for topic, topic_def in m5oc.items():
            topic_def["designs"] = self.filter_designs(topic_def["designs"])

        return {
            topic: topic_def
            for topic, topic_def in m5oc.items()
            if topic_def["designs"]
        }


@reposBP.route("/", methods=["GET"])
def index():
    project = Project.find()
    onlyfiles = [f for f in project.model_dir().iterdir() if f.is_file()]

    path = project.model_dir()
    dashboardsParser = M5oCollectionParser(path, M5oCollectionParserTypes.Dashboard)
    reportsParser = M5oCollectionParser(path, M5oCollectionParserTypes.Report)

    reportsFiles = reportsParser.parse()
    dashboardFiles = dashboardsParser.parse()

    sortedM5oFiles = {
        "dashboards": {"label": "Dashboards", "items": dashboardFiles},
        "documents": {"label": "Documents", "items": []},
        "topics": {"label": "Topics", "items": []},
        "reports": {"label": "Reports", "items": reportsFiles},
        "tables": {"label": "Tables", "items": []},
    }
    onlydocs = project.model_dir().parent.glob("*.md")

    for d in onlydocs:
        file_dict = MeltanoAnalysisFileParser.fill_base_m5o_dict(
            d.relative_to(project.root), str(d.name)
        )
        sortedM5oFiles["documents"]["items"].append(file_dict)

    for f in onlyfiles:
        filename, ext = os.path.splitext(f)
        if ext != ".m5o":
            continue

        # filename splittext twice occurs due to current *.type.extension convention (two dots)
        filename = filename.lower()
        filename, ext = os.path.splitext(filename)
        file_dict = MeltanoAnalysisFileParser.fill_base_m5o_dict(
            f.relative_to(project.root), filename
        )
        if ext == ".topic":
            sortedM5oFiles["topics"]["items"].append(file_dict)
        if ext == ".table":
            sortedM5oFiles["tables"]["items"].append(file_dict)

    m5o_parser = MeltanoAnalysisFileParser(project)

    for package in m5o_parser.packages():
        package_files = MeltanoAnalysisFileParser.package_files(package)
        sortedM5oFiles["topics"]["items"] += package_files["topics"]["items"]
        sortedM5oFiles["tables"]["items"] += package_files["tables"]["items"]

    if not len(sortedM5oFiles["topics"]["items"]):
        return jsonify(
            {
                "result": False,
                "errors": [{"message": "Missing topic file(s)", "file_name": "*"}],
                "files": [],
            }
        )

    return jsonify(sortedM5oFiles)


def lint_all(compile):
    project = Project.find()
    compiler = ProjectCompiler(project)
    try:
        compiler.parse()
    except MeltanoAnalysisFileParserError as e:
        return jsonify(
            {
                "result": False,
                "errors": [{"message": e.message, "file_name": e.file_name}],
            }
        )
    if compile:
        compiler.compile()
    return jsonify({"result": True})


@reposBP.errorhandler(MeltanoAnalysisFileParserError)
def handle_meltano_analysis_file_parser_error(e):
    return (
        jsonify(
            {
                "result": False,
                "errors": [{"message": e.message, "file_name": e.file_name}],
            }
        ),
        500,
    )


@reposBP.errorhandler(FileNotFoundError)
def handle_file_not_found(e):
    return jsonify({"result": False, "error": str(e)}), 404


@reposBP.route("/lint", methods=["GET"])
def lint():
    return lint_all(False)


@reposBP.route("/sync", methods=["GET"])
def sync():
    return lint_all(True)


@reposBP.route("/designs/<path:namespace>/<topic_name>/<design_name>", methods=["GET"])
def model_design(namespace, topic_name, design_name):
    repos_helper = ReposHelper()
    topic = repos_helper.get_topic(namespace, topic_name)

    designs = topic["designs"]
    design = next(e for e in designs if e["name"] == design_name)

    return jsonify(design)


@reposBP.route("/models", methods=["GET"])
def model_index():
    project = Project.find()
    topicsFile = project.run_dir("models", "topics.index.m5oc")
    path = Path(topicsFile)

    topics = json.load(open(path)) if path.is_file() else {}

    # for now let's freeze the keys, so the we can re-use the
    # key name as the design's name, but this should probably
    # return a list
    topics = freeze_keys(topics)
    return jsonify(topics)


@reposBP.route("/<path:namespace>/<topic_name>", methods=["GET"])
def model_topic(namespace, topic_name):
    repos_helper = ReposHelper()
    topic = repos_helper.get_topic(namespace, topic_name)

    return jsonify(topic)
