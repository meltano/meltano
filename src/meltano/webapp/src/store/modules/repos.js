import lodash from 'lodash'

import repoApi from '@/api/repo'
import utils from '@/utils/utils'

const defaultState = utils.deepFreeze({
  activeView: {
    file: '',
    is_markdown: false,
    populated: false,
  },
  errors: [],
  files: {},
  loadingUpdate: false,
  loadingValidation: false,
  models: {},
  validated: false,
})

const getters = {
  hasCode(state) {
    return state.activeView.populated && !state.activeView.is_markdown
  },

  hasError(state) {
    return state.errors && state.errors.length
  },

  hasFiles(state) {
    return state.files.hasOwnProperty('topics') && state.files.topics.items
  },

  hasMarkdown(state) {
    return state.activeView.populated && state.activeView.is_markdown
  },

  hasModels(state) {
    return !lodash.isEmpty(state.models)
  },

  passedValidation(state) {
    return state.validated && state.errors && !state.errors.length
  },

  urlForModelDesign: () => (model, design) => `/analyze/${model}/${design}`,
}

const actions = {
  getModels({ commit }) {
    repoApi.models().then((response) => {
      commit('setModels', response.data)
    })
  },

  getRepo({ state, commit }) {
    repoApi.index().then((response) => {
      const files = response.data
      commit('setValidatedState', response.data)
      state.loadingValidation = false
      commit('setRepoFiles', { files })
    })
  },

  lint({ state, commit }) {
    state.loadingValidation = true
    repoApi
      .lint()
      .then((response) => {
        commit('setValidatedState', response.data)
        state.loadingValidation = false
      })
      .catch(() => {
        state.loadingValidation = false
      })
  },

  sync({ state, commit, dispatch }) {
    state.loadingUpdate = true
    repoApi
      .sync()
      .then((response) => {
        dispatch('getModels')
        commit('setValidatedState', response.data)
        state.loadingUpdate = false
      })
      .catch(() => {
        state.loadingUpdate = false
      })
  },
}

const mutations = {
  setCurrentFileTable(state, file) {
    state.activeView = file
  },

  setModels(state, models) {
    state.models = models
  },

  setRepoFiles(state, { files }) {
    state.files = files
  },

  setValidatedState(state, validated) {
    state.errors = []
    state.validated = true
    // validation failed, so there will be errors
    if (!validated.result) {
      state.errors = validated.errors
    }
  },
}

export default {
  namespaced: true,
  state: lodash.cloneDeep(defaultState),
  getters,
  actions,
  mutations,
}
