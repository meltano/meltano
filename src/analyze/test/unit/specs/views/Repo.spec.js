import { mount, createLocalVue, shallowMount } from '@vue/test-utils';
import Vuex from 'vuex';
import Repo from '@/views/Repo';
import repos from '@/store/modules/repos';

const localVue = createLocalVue();
localVue.use(Vuex);

describe('Repo.vue', () => {
  let actions;
  let state;
  let store;

  beforeEach(() => {
    state = repos.state;
    actions = {
      getRepo: jest.fn(),
      getFile: jest.fn(),
      sync: jest.fn(),
    };
    store = new Vuex.Store({
      modules: {
        repos: {
          namespaced: true,
          state,
          actions,
          getters: repos.getters,
        },
      },
    });
  });

  it('calls getRepo() and sync() via created() lifecycle hook', () => {
    const wrapper = shallowMount(Repo, { store, localVue });

    expect(wrapper.html()).toBeTruthy();
    expect(actions.getRepo).toHaveBeenCalled();
    expect(actions.sync).toHaveBeenCalled();
  });

  it('renders no code or markdown by default', () => {
    const wrapper = mount(Repo, { store, localVue });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders markdown in the preview pane for markdown files', () => {
    state.activeView = { is_markdown: true, file: '<h1>Title</h1>', populated: true };
    const wrapper = mount(Repo, { store, localVue });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders code in the preview pane for code files', () => {
    state.activeView = { is_markdown: false, file: '{ "title": "Title" }', populated: true };
    const wrapper = mount(Repo, { store, localVue });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders the dashboard list when dashboards exist and calls getFile() when a dashboard is clicked', () => {
    state.files = {
      dashboards: {
        label: 'Dashboards',
        items: [{
          createdAt: 1551459531.431577,
          description: '',
          id: 'F5KXGZLSOMXWI23ON54C2Z3JORWGCYRPIRXWG5LNMVXHI4ZPKBZG62TFMN2HGL3DMFZGE33OFVTWS5DGNRUXQL3NN5SGK3BPMIXGIYLTNBRG6YLSMQXG2NLP',
          name: 'Some',
          path: 'some.dashboard.m5o',
          reportIds: [],
          slug: 'some',
          version: '1.0.0',
        }],
      },
    };
    const wrapper = mount(Repo, { store, localVue });

    expect(wrapper.element).toMatchSnapshot();

    wrapper.find('.js-dashboards-some').trigger('click');

    expect(actions.getFile).toHaveBeenCalled();
  });

  it('renders the document list when documents exist and calls getFile() when a document is clicked', () => {
    state.files = {
      documents: {
        label: 'Documents',
        items: [{
          createdAt: 1551990614.693281,
          id: 'F5KXGZLSOMXWI23ON54C2Z3JORWGCYRPIRXWG5LNMVXHI4ZPKBZG62TFMN2HGL3DMFZGE33OFVTWS5DGNRUXQL2SIVAUITKFFZWWI===',
          name: 'README',
          path: 'README.md',
          slug: 'readme-md',
        }],
      },
    };
    const wrapper = mount(Repo, { store, localVue });

    expect(wrapper.element).toMatchSnapshot();

    wrapper.find('.js-documents-readme').trigger('click');

    expect(actions.getFile).toHaveBeenCalled();
  });

  it('renders the report list when reports exist and calls getFile() when a report is clicked', () => {
    state.files = {
      reports: {
        label: 'Reports',
        items: [{
          chartType: 'BarChart',
          createdAt: 1551461064.508327,
          design: 'region',
          id: 'F5KXGZLSOMXWI23ON54C2Z3JORWGCYRPIRXWG5LNMVXHI4ZPKBZG62TFMN2HGL3DMFZGE33OFVTWS5DGNRUXQL3NN5SGK3BPMMZC44TFOBXXE5BONU2W6===',
          model: 'carbon',
          name: 'Some',
          path: 'some.report.m5o',
          queryPayload: {},
          slug: 'some',
          version: '1.0.0',
        }],
      },
    };
    const wrapper = mount(Repo, { store, localVue });

    expect(wrapper.element).toMatchSnapshot();

    wrapper.find('.js-reports-some').trigger('click');

    expect(actions.getFile).toHaveBeenCalled();
  });

  it('renders the table list when tables exist and calls getFile() when a table is clicked', () => {
    state.files = {
      tables: {
        label: 'Tables',
        items: [{
          createdAt: 1551989728.153954,
          id: 'MVYGS43PMRSXGLTUMFRGYZJONU2W6===',
          name: 'episodes',
          path: 'episodes.table.m5o',
          slug: 'episodes',
        }],
      },
    };
    const wrapper = mount(Repo, { store, localVue });

    expect(wrapper.element).toMatchSnapshot();

    wrapper.find('.js-tables-episodes').trigger('click');

    expect(actions.getFile).toHaveBeenCalled();
  });

  it('renders the topic list when topics exist and calls getFile() when a topic is clicked', () => {
    state.files = {
      topics: {
        label: 'Topics',
        items: [{
          version: 1,
          name: 'carbon',
          connection: 'runners_db',
          label: 'carbon intensity',
          designs: {},
        }],
      },
    };
    const wrapper = mount(Repo, { store, localVue });

    expect(wrapper.element).toMatchSnapshot();

    wrapper.find('.js-topics-carbon').trigger('click');

    expect(actions.getFile).toHaveBeenCalled();
  });
});
