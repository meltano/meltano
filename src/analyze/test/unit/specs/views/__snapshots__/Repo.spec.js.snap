// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repo.vue renders code in the preview pane for code files 1`] = `
<div
  class="columns is-gapless"
>
  <aside
    class="column is-one-quarter vh-scrollable has-background-white-bis"
  >
    <div
      class="box is-radiusless is-shadowless"
    >
      <div>
        <div
          class="level"
        >
          <div
            class="level-left"
          >
            <div
              class="field has-addons"
            >
              <div
                class="control"
              >
                <a
                  class="button is-small"
                  href="#"
                >
                  Lint
                </a>
              </div>
               
              <div
                class="control"
              >
                <a
                  class="button is-small"
                  href="#"
                >
                  Sync
                </a>
              </div>
            </div>
          </div>
           
          <div
            class="level-right"
          >
            <!---->
             
            <span
              class="tag is-warning"
            >
              Unvalidated
            </span>
             
            <!---->
          </div>
        </div>
         
        <!---->
         
        <p
          class="menu-label"
        >
          
        No files found
      
        </p>
         
      </div>
    </div>
  </aside>
   
  <section
    class="column is-three-quarters vh-scrollable"
  >
    <div
      class="box is-radiusless is-shadowless"
    >
      <div>
        <!---->
         
        <div
          class="js-code-preview is-paddingless code-container"
        >
          <div
            class="has-background-white"
          >
            <pre>
              {
  "title": "Title"
}
            </pre>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`Repo.vue renders markdown in the preview pane for markdown files 1`] = `
<div
  class="columns is-gapless"
>
  <aside
    class="column is-one-quarter vh-scrollable has-background-white-bis"
  >
    <div
      class="box is-radiusless is-shadowless"
    >
      <div>
        <div
          class="level"
        >
          <div
            class="level-left"
          >
            <div
              class="field has-addons"
            >
              <div
                class="control"
              >
                <a
                  class="button is-small"
                  href="#"
                >
                  Lint
                </a>
              </div>
               
              <div
                class="control"
              >
                <a
                  class="button is-small"
                  href="#"
                >
                  Sync
                </a>
              </div>
            </div>
          </div>
           
          <div
            class="level-right"
          >
            <!---->
             
            <span
              class="tag is-warning"
            >
              Unvalidated
            </span>
             
            <!---->
          </div>
        </div>
         
        <!---->
         
        <p
          class="menu-label"
        >
          
        No files found
      
        </p>
         
      </div>
    </div>
  </aside>
   
  <section
    class="column is-three-quarters vh-scrollable"
  >
    <div
      class="box is-radiusless is-shadowless"
    >
      <div>
        <!---->
         
        <div>
          <div
            class="js-markdown-preview has-background-white"
          >
            <h1>
              Title
            </h1>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`Repo.vue renders no code or markdown by default 1`] = `
<div
  class="columns is-gapless"
>
  <aside
    class="column is-one-quarter vh-scrollable has-background-white-bis"
  >
    <div
      class="box is-radiusless is-shadowless"
    >
      <div>
        <div
          class="level"
        >
          <div
            class="level-left"
          >
            <div
              class="field has-addons"
            >
              <div
                class="control"
              >
                <a
                  class="button is-small"
                  href="#"
                >
                  Lint
                </a>
              </div>
               
              <div
                class="control"
              >
                <a
                  class="button is-small"
                  href="#"
                >
                  Sync
                </a>
              </div>
            </div>
          </div>
           
          <div
            class="level-right"
          >
            <!---->
             
            <span
              class="tag is-warning"
            >
              Unvalidated
            </span>
             
            <!---->
          </div>
        </div>
         
        <!---->
         
        <p
          class="menu-label"
        >
          
        No files found
      
        </p>
         
      </div>
    </div>
  </aside>
   
  <section
    class="column is-three-quarters vh-scrollable"
  >
    <div
      class="box is-radiusless is-shadowless"
    >
      <div>
        <div>
          <div
            class="empty-state
        has-background-white
        has-text-centered
        is-size-4
        is-uppercase"
          >
            
        Select a file
      
          </div>
        </div>
         
        <!---->
      </div>
    </div>
  </section>
</div>
`;

exports[`Repo.vue renders the dashboard list when dashboards exist and calls getFile() when a dashboard is clicked 1`] = `
<div
  class="columns is-gapless"
>
  <aside
    class="column is-one-quarter vh-scrollable has-background-white-bis"
  >
    <div
      class="box is-radiusless is-shadowless"
    >
      <div>
        <div
          class="level"
        >
          <div
            class="level-left"
          >
            <div
              class="field has-addons"
            >
              <div
                class="control"
              >
                <a
                  class="button is-small"
                  href="#"
                >
                  Lint
                </a>
              </div>
               
              <div
                class="control"
              >
                <a
                  class="button is-small"
                  href="#"
                >
                  Sync
                </a>
              </div>
            </div>
          </div>
           
          <div
            class="level-right"
          >
            <!---->
             
            <span
              class="tag is-warning"
            >
              Unvalidated
            </span>
             
            <!---->
          </div>
        </div>
         
        <!---->
         
        <p
          class="menu-label"
        >
          
        No files found
      
        </p>
         
        <p
          class="menu-label"
        >
          
        Dashboards
      
        </p>
         
        <ul
          class="menu-list"
        >
          <li>
            <div
              class="columns is-vcentered"
            >
              <div
                class="column"
              >
                <a
                  class="js-dashboards-some"
                >
                  
                  Some
                
                </a>
              </div>
               
              <div
                class="column is-one-fifth"
              >
                <router-link
                  class="button is-secondary is-light is-small is-pulled-right"
                  to="[object Object]"
                >
                  <font-awesome-icon
                    icon="arrow-right"
                  />
                </router-link>
              </div>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </aside>
   
  <section
    class="column is-three-quarters vh-scrollable"
  >
    <div
      class="box is-radiusless is-shadowless"
    >
      <div>
        <!---->
         
        <div
          class="js-code-preview is-paddingless code-container"
        >
          <div
            class="has-background-white"
          >
            <pre>
              {
  "title": "Title"
}
            </pre>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`Repo.vue renders the document list when documents exist and calls getFile() when a document is clicked 1`] = `
<div
  class="columns is-gapless"
>
  <aside
    class="column is-one-quarter vh-scrollable has-background-white-bis"
  >
    <div
      class="box is-radiusless is-shadowless"
    >
      <div>
        <div
          class="level"
        >
          <div
            class="level-left"
          >
            <div
              class="field has-addons"
            >
              <div
                class="control"
              >
                <a
                  class="button is-small"
                  href="#"
                >
                  Lint
                </a>
              </div>
               
              <div
                class="control"
              >
                <a
                  class="button is-small"
                  href="#"
                >
                  Sync
                </a>
              </div>
            </div>
          </div>
           
          <div
            class="level-right"
          >
            <!---->
             
            <span
              class="tag is-warning"
            >
              Unvalidated
            </span>
             
            <!---->
          </div>
        </div>
         
        <!---->
         
        <p
          class="menu-label"
        >
          
        No files found
      
        </p>
         
        <p
          class="menu-label"
        >
          
        Documents
      
        </p>
         
        <ul
          class="menu-list"
        >
          <li>
            <div
              class="columns is-vcentered"
            >
              <div
                class="column"
              >
                <a
                  class="js-documents-readme"
                >
                  
                  README
                
                </a>
              </div>
               
              <!---->
            </div>
          </li>
        </ul>
      </div>
    </div>
  </aside>
   
  <section
    class="column is-three-quarters vh-scrollable"
  >
    <div
      class="box is-radiusless is-shadowless"
    >
      <div>
        <!---->
         
        <div
          class="js-code-preview is-paddingless code-container"
        >
          <div
            class="has-background-white"
          >
            <pre>
              {
  "title": "Title"
}
            </pre>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`Repo.vue renders the report list when reports exist and calls getFile() when a report is clicked 1`] = `
<div
  class="columns is-gapless"
>
  <aside
    class="column is-one-quarter vh-scrollable has-background-white-bis"
  >
    <div
      class="box is-radiusless is-shadowless"
    >
      <div>
        <div
          class="level"
        >
          <div
            class="level-left"
          >
            <div
              class="field has-addons"
            >
              <div
                class="control"
              >
                <a
                  class="button is-small"
                  href="#"
                >
                  Lint
                </a>
              </div>
               
              <div
                class="control"
              >
                <a
                  class="button is-small"
                  href="#"
                >
                  Sync
                </a>
              </div>
            </div>
          </div>
           
          <div
            class="level-right"
          >
            <!---->
             
            <span
              class="tag is-warning"
            >
              Unvalidated
            </span>
             
            <!---->
          </div>
        </div>
         
        <!---->
         
        <p
          class="menu-label"
        >
          
        No files found
      
        </p>
         
        <p
          class="menu-label"
        >
          
        Reports
      
        </p>
         
        <ul
          class="menu-list"
        >
          <li>
            <div
              class="columns is-vcentered"
            >
              <div
                class="column"
              >
                <a
                  class="js-reports-some"
                >
                  
                  Some
                
                </a>
              </div>
               
              <div
                class="column is-one-fifth"
              >
                <router-link
                  class="button is-secondary is-light is-small is-pulled-right"
                  to="[object Object]"
                >
                  <font-awesome-icon
                    icon="arrow-right"
                  />
                </router-link>
              </div>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </aside>
   
  <section
    class="column is-three-quarters vh-scrollable"
  >
    <div
      class="box is-radiusless is-shadowless"
    >
      <div>
        <!---->
         
        <div
          class="js-code-preview is-paddingless code-container"
        >
          <div
            class="has-background-white"
          >
            <pre>
              {
  "title": "Title"
}
            </pre>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`Repo.vue renders the table list when tables exist and calls getFile() when a table is clicked 1`] = `
<div
  class="columns is-gapless"
>
  <aside
    class="column is-one-quarter vh-scrollable has-background-white-bis"
  >
    <div
      class="box is-radiusless is-shadowless"
    >
      <div>
        <div
          class="level"
        >
          <div
            class="level-left"
          >
            <div
              class="field has-addons"
            >
              <div
                class="control"
              >
                <a
                  class="button is-small"
                  href="#"
                >
                  Lint
                </a>
              </div>
               
              <div
                class="control"
              >
                <a
                  class="button is-small"
                  href="#"
                >
                  Sync
                </a>
              </div>
            </div>
          </div>
           
          <div
            class="level-right"
          >
            <!---->
             
            <span
              class="tag is-warning"
            >
              Unvalidated
            </span>
             
            <!---->
          </div>
        </div>
         
        <!---->
         
        <p
          class="menu-label"
        >
          
        No files found
      
        </p>
         
        <p
          class="menu-label"
        >
          
        Tables
      
        </p>
         
        <ul
          class="menu-list"
        >
          <li>
            <div
              class="columns is-vcentered"
            >
              <div
                class="column"
              >
                <a
                  class="js-tables-episodes"
                >
                  
                  episodes
                
                </a>
              </div>
               
              <!---->
            </div>
          </li>
        </ul>
      </div>
    </div>
  </aside>
   
  <section
    class="column is-three-quarters vh-scrollable"
  >
    <div
      class="box is-radiusless is-shadowless"
    >
      <div>
        <!---->
         
        <div
          class="js-code-preview is-paddingless code-container"
        >
          <div
            class="has-background-white"
          >
            <pre>
              {
  "title": "Title"
}
            </pre>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`Repo.vue renders the topic list when topics exist and calls getFile() when a topic is clicked 1`] = `
<div
  class="columns is-gapless"
>
  <aside
    class="column is-one-quarter vh-scrollable has-background-white-bis"
  >
    <div
      class="box is-radiusless is-shadowless"
    >
      <div>
        <div
          class="level"
        >
          <div
            class="level-left"
          >
            <div
              class="field has-addons"
            >
              <div
                class="control"
              >
                <a
                  class="button is-small"
                  href="#"
                >
                  Lint
                </a>
              </div>
               
              <div
                class="control"
              >
                <a
                  class="button is-small"
                  href="#"
                >
                  Sync
                </a>
              </div>
            </div>
          </div>
           
          <div
            class="level-right"
          >
            <!---->
             
            <span
              class="tag is-warning"
            >
              Unvalidated
            </span>
             
            <!---->
          </div>
        </div>
         
        <!---->
         
        <!---->
         
        <p
          class="menu-label"
        >
          
        Topics
      
        </p>
         
        <ul
          class="menu-list"
        >
          <li>
            <div
              class="columns is-vcentered"
            >
              <div
                class="column"
              >
                <a
                  class="is-active js-topics-carbon"
                >
                  
                  carbon
                
                </a>
              </div>
               
              <!---->
            </div>
          </li>
        </ul>
      </div>
    </div>
  </aside>
   
  <section
    class="column is-three-quarters vh-scrollable"
  >
    <div
      class="box is-radiusless is-shadowless"
    >
      <div>
        <!---->
         
        <div
          class="js-code-preview is-paddingless code-container"
        >
          <div
            class="has-background-white"
          >
            <pre>
              {
  "title": "Title"
}
            </pre>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
